File: mobile/modules/dispo-legacy-preferences/ios/DispoLegacyPreferencesModule.swift
import ExpoModulesCore

private struct LegacyPreferencesResult: Record {
  @Field var groupLastSeen: [String: String]?
  @Field var language: String?
  @Field var notificationsEnabled: Bool?
  @Field var openedGigIds: [String]?
  @Field var pushGroups: Bool?
  @Field var pushMessages: Bool?
  @Field var pushSos: Bool?
  @Field var schoolLastSeen: [String: String]?
  @Field var sosShowAll: Bool?
  @Field var theme: String?
}

public final class DispoLegacyPreferencesModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DispoLegacyPreferences")

    Function("readAsync") { () -> LegacyPreferencesResult in
      let values = LegacyPreferencesReader.read(UserDefaults.standard)
      let result = LegacyPreferencesResult()
      result.groupLastSeen = values.groupLastSeen
      result.language = values.language
      result.notificationsEnabled = values.notificationsEnabled
      result.openedGigIds = values.openedGigIds
      result.pushGroups = values.pushGroups
      result.pushMessages = values.pushMessages
      result.pushSos = values.pushSos
      result.schoolLastSeen = values.schoolLastSeen
      result.sosShowAll = values.sosShowAll
      result.theme = values.theme
      return result
    }
  }
}

File: mobile/modules/dispo-legacy-preferences/android/src/main/java/ch/dispo/modules/legacypreferences/DispoLegacyPreferencesModule.kt
package ch.dispo.modules.legacypreferences

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DispoLegacyPreferencesModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DispoLegacyPreferences")

    Function("readAsync") {
      val context = requireNotNull(appContext.reactContext)
      val preferences = context.getSharedPreferences(
        LegacyPreferencesReader.PREFERENCES_FILE,
        Context.MODE_PRIVATE,
      )
      val values = LegacyPreferencesReader.read(preferences.all)
      mapOf(
        "groupLastSeen" to values.groupLastSeen,
        "groupLastSeenByProfile" to values.groupLastSeenByProfile,
        "language" to values.language,
        "notificationsEnabled" to values.notificationsEnabled,
        "openedGigIds" to values.openedGigIds,
        "pushGroups" to values.pushGroups,
        "pushMessages" to values.pushMessages,
        "pushSos" to values.pushSos,
        "schoolLastSeen" to values.schoolLastSeen,
        "sosShowAll" to null,
        "theme" to values.theme,
      )
    }

    Function("readSupabaseSessionAsync") {
      val context = requireNotNull(appContext.reactContext)
      val preferences = context.getSharedPreferences(
        LegacyPreferencesReader.SUPABASE_PREFERENCES_FILE,
        Context.MODE_PRIVATE,
      )
      LegacyPreferencesReader.readSupabaseSession(preferences.all)
    }
  }
}
